src/main/scala/is/hail/io/bgen/BgenRDD.scala

```diff
@@ -0,0 +1,335 @@
+  rsid: Boolean,
+  fileRowIndex: Boolean
+)
+
+private case class BgenSettings (
+  nSamples: Int,
+  nVariants: Int,
+  entries: EntriesSetting,
+  rowFields: RowFields,
+  rg: Option[ReferenceGenome],
+  private val userContigRecoding: Map[String, String],
+  skipInvalidLoci: Boolean
+) {
+  private[this] val typedRowFields = Array(
+    (true, "locus" -> TLocus.schemaFromRG(rg)),
+    (true, "alleles" -> TArray(TString())),
+    (rowFields.rsid, "rsid" -> TString()),
+    (rowFields.varid, "varid" -> TString()),
+    (rowFields.fileRowIndex, "file_row_idx" -> TInt64()))
+    .withFilter(_._1).map(_._2)
+
+  private[this] val typedEntryFields: Array[(String, Type)] = entries match {
+    case NoEntries => Array.empty
+    case EntriesWithFields(gt, gp, dosage) => Array(
+      (gt, "GT" -> TCall()),
+      (gp, "GP" -> +TArray(+TFloat64())),
+      (dosage, "dosage" -> +TFloat64()))
+        .withFilter(_._1).map(_._2)
+  }
+
+  val matrixType: MatrixType = MatrixType.fromParts(
+    globalType = TStruct.empty(),
+    colKey = Array("s"),
+    colType = TStruct("s" -> TString()),
+    rowType = TStruct(typedRowFields: _*),
+    rowKey = Array("locus", "alleles"),
+    rowPartitionKey = Array("locus"),
+    entryType = TStruct(typedEntryFields: _*))
+
+  val typ: TStruct = entries match {
+    case NoEntries =>
+      matrixType.rowType
+    case _: EntriesWithFields =>
+      matrixType.rvRowType
+  }
+
+  def recodeContig(bgenContig: String): String = {
+    val hailContig = bgenContig match {
+      case "23" => "X"
+      case "24" => "Y"
+      case "25" => "X"
+      case "26" => "MT"
+      case x => x
+    }
+    userContigRecoding.getOrElse(hailContig, hailContig)
+  }
+}
+
+object BgenRDD {
+  def apply(
+    sc: SparkContext,
+    files: Seq[BgenHeader],
+    minPartitions: Option[Int],
+    includedVariantsPerFile: Map[String, Seq[Int]],
+    settings: BgenSettings
+  ): ContextRDD[RVDContext, RegionValue] =
+    ContextRDD(
+      new BgenRDD(sc, files, minPartitions, includedVariantsPerFile, settings))
+}
+
+private class BgenRDD(
+  sc: SparkContext,
+  files: Seq[BgenHeader],
+  minPartitions: Option[Int],
+  includedVariantsPerFile: Map[String, Seq[Int]],
+  settings: BgenSettings
+) extends RDD[RVDContext => Iterator[RegionValue]](sc, Nil) {
+  private[this] val defaultMinPartitions =
+    sc.defaultMinPartitions
+  private[this] val parts = BgenRDDPartitions(
+    sc,
+    files,
+    minPartitions.getOrElse(defaultMinPartitions),
+    includedVariantsPerFile,
+    settings)
+
+  protected def getPartitions: Array[Partition] = parts
+
+  def compute(split: Partition, context: TaskContext): Iterator[RVDContext => Iterator[RegionValue]] =
+    Iterator.single { (ctx: RVDContext) =>
+      new BgenRecordIterator(ctx, split.asInstanceOf[BgenPartition], settings).flatten }
+}
+
+private class BgenRecordIterator(
+  ctx: RVDContext,
+  p: BgenPartition,
+  settings: BgenSettings
+) extends Iterator[Option[RegionValue]] {
```

Comment:
Won't an iterator of `Option`s do a lot of allocations? It doesn't look that hard to write `next` to skip over invalid loci internally instead of returning `None`, and I would argue that's clearer too, since the only user of `BgenRecordIterator` is just filtering out the `None`s anyways.